File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses);
        vector<int>indeg(numCourses, 0);
        for(auto pre:prerequisites){
            adj[pre[1]].push_back(pre[0]);
            indeg[pre[0]]++;
        }
        queue<int>q;
        for(int i= 0; i<numCourses;i++){
            if(indeg[i] == 0)q.push(i);
        }
        int c = 0;
        while(!q.empty()){
            int n = q.front();
            q.pop();
            c++;
            for(int x: adj[n]){
                indeg[x]--;
                if(indeg[x] == 0){
                    q.push(x);
                }
            }
        }
        if(c == numCourses)return true;
        return false;
    }
};

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        ArrayList<ArrayList<Integer>>adj = new ArrayList<>();
        int [] indeg = new int[numCourses];
        for(int i = 0; i < numCourses; i++){
            adj.add(new ArrayList<>());
            indeg[i] = 0;
        }
        for(int [] edj : prerequisites){
            int ai = edj[0];
            int bi = edj[1];
            adj.get(bi).add(ai);
            indeg[ai]++;
        }
        Queue<Integer> q = new LinkedList<>();
        for(int i = 0; i< numCourses ;i++){
            if(indeg[i] == 0){
                q.offer(i);
            }
        }
        int cnt = 0;
        while(!q.isEmpty()){
            int node = q.poll();
            cnt++;
            for(int a: adj.get(node)){
                indeg[a]--;
                if(indeg[a]== 0){
                    q.offer(a);
                }
            }
        }
        if(cnt == numCourses) return true;
        return false;
    }
}